Compile less executables in tests

alexcrichton opened the pull request:
Refactor a bit with more aggressive caching to compile at most one Rust program when testing.

alexcrichton committed:
Refactor tests with less tuples and more objects

In prepration for the next commit...

alexcrichton committed:
Only compile executables once per process

Use auxiliary files to store version information instead of embedding it
directly in the executable.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 707ca86 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 707ca864c3a5e57cac706cafb2d4714879536ada -->

bors commented:
:hourglass: Testing commit 707ca864c3a5e57cac706cafb2d4714879536ada with merge 622d6a91b5d9da8ca5b9618b44ed32aaf22f0398...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang-nursery/rustup.rs/builds/263629645?utm_source=github_status&utm_medium=notification)

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit f976a22 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton f976a224297afc2565c8cb34aa6f8bacb709c06d -->

bors commented:
:hourglass: Testing commit f976a224297afc2565c8cb34aa6f8bacb709c06d with merge 8a4dacc896cbcc52b5c16c8192d4d0de2f4e51f3...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang-nursery/rustup.rs/builds/263740823?utm_source=github_status&utm_medium=notification)

alexcrichton committed:
Update a slew of dependencies

alexcrichton committed:
Only build one matrix on AppVeyor for PRs

Skip most builds as they end up just clogging the queue. Let's just take one
passing builder and test the full suite on the `auto` branch.

alexcrichton committed:
Switch travis to rustup and one build per PR

Like AppVeyor this only schedules one build to run on PRs, and everything else
will exit quickly. This'll help keep the queue clear for PRs when they come
through. Also switch Travis to using rustup instead of `rustup.sh` manually.

alexcrichton committed:
Move off the old cross builder where possible

Many of the toolchains in the old cross builder came from upstream Ubuntu
packages, so just use those but in their own separate container we assemble
locally. No need to download multi-gigabyte docker images when all we need is
one toolchain!

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 761d351 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 761d351320980c14127138d6192f82f531542778 -->

bors commented:
:hourglass: Testing commit 761d351320980c14127138d6192f82f531542778 with merge b021a9da823eab03f13d4cf4f3743a55da5c0502...

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit 761d351320980c14127138d6192f82f531542778 with merge ee227d89b6ce663a0cb30ef1da2e9c3f3e45c365...

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 65aea18 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 65aea180c5a37a6d77cb3ce0100e5775f2d98407 -->

bors commented:
:hourglass: Testing commit 65aea180c5a37a6d77cb3ce0100e5775f2d98407 with merge e04b84ef318e57c3ab678af3f2a6f64906fcdab6...

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 3e99a69 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 3e99a693a3b811bb1d8098dfa254f1bb543b6e05 -->

bors commented:
:hourglass: Testing commit 3e99a693a3b811bb1d8098dfa254f1bb543b6e05 with merge 76fecc53443e1657ed7b1e2d89fdaa9d652d6b88...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/brson/rustup-rs/build/1.0.1358), [status-travis](https://travis-ci.org/rust-lang-nursery/rustup.rs/builds/263864216?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing 76fecc53443e1657ed7b1e2d89fdaa9d652d6b88 to master...
